### DISCUSSION TITLE
Best way to install on a disk station in the Docker environment (called Container Manager here)

### DISCUSSION THREAD

**Discussion by megabyte0469** [s-allius/tsun-gen3-proxy#16] [Q&A] [answered]
Hi would like to deploy the whole thing here. Have 2 TSUN 3rd generation inverters. 
Would then like to include this in my Home Assistant environment.
What must be active in the Home Assistant environment?
Unfortunately the whole thing refuses to run in the "Synology Docker" environment. Is there anything special to be aware of when it is started ?! Thanks for an answer and the help.

**Comment by s-allius** [comment]
Hello megabyte0469,
I myself also run my stack on a Synolgy NAS. With today's release 0.4.0 the operation of multiple inverters on one proxy is also registered correctly in Home Assistant. For an installation guide I had not yet found time. But don't worry we should get that right.

Since I have not yet fully understood the problem, first a rough overview
- in docker you need to install a MQTT broker and home assistant in addition to the proxy. The easiest way to do this is with docker compose. For this you can use the 'container manager' from Synology, or like me, set up portainer. The advantage of this approach is that you can define dependencies at startup.
- for each container you have to create a directory on the NAS, where the containers can store their persistent data.
- In the next step you have to create an account in the MQTT proxy, or you start without credentials, which is supported by Home Assistant and my proxy.
- After that you can install the MQTT integration in the Home Assistant and configure the account for the MQTT broker.
- Now only the configuration for the proxy has to be adjusted. After the first start of the proxy container it copies a 'config.example.toml' file into the config directory on the NAS. Simply rename this file to 'config.toml' and enter the access data for the MQTT broker in it, if required.
- It is very important to change the DNS resolution from 'logger.talent-monitoring.com' to the IP address of the NAS in your router!

```
After starting the proxy, you should see in the docker console something like:
######################################################
# prepare: 'tsun-proxy' Version:0.4.0
# for running with UserID:1000, GroupID:1000
#
######################################################
#
'2023-10-16 22:32:25  INFO | root | Server "tsun-proxy - 0.4.0" will be started'
'2023-10-16 22:32:25  INFO | mqtt | start MQTT: host:mqtt  port:1883  user:?????'

```
If everything has worked so far, you can test the installation as follows:
- restart the home assistant. This sends two messages to the MQTT broker, which will be reported by the proxy.  In the proxy console should appear these two lines:
```
'2023-10-16 23:36:15  INFO | mqtt | Home-Assistant Status: offline'
'2023-10-16 23:36:34  INFO | mqtt | Home-Assistant Status: online'
```
- now load this site in your browser 'http://logger.talent-monitoring.com:5005', so that the browser tries to establish a tcp connection to the proxy. In the proxy console you should see:
```'2023-10-16 23:38:42  INFO | root | Accept connection from  ('<ip-addr>', <port>)'```

Of course, the browser can't load anything, which is expected. When you close the browser window, the connection will be closed and you should see, this in the console:
```
'2023-10-16 23:39:20 WARNING | conn | In loop for ('<ip-addr>', <port>): Peer closed.'
'2023-10-16 23:39:20  INFO | root | Server loop stopped for ('<ip-addr>', <port>)'
```
Now you system is working and you must restart your inverter., so that he request the DNS resolution again. For this simply disconnect the panels, redo the pairing or wait until the next day. Than you should see all the entities in the home assistant overview dashboard.

If anything up to here has not worked, please ask me again. then i will be happy to help with even more details.

Cheers
Stefan

**Reply by megabyte0469** [reply]
Hallo Stefan,
diesmal in Deutsch. ;-)
Danke für die Anleitung. Bin erst jetzt dazu gekommen die Antwort zu lesen....
Ich hab meinen Home Assistent auf einen extra Rechner laufen. Nicht auf der Diskstation. 
Auf dem Home Assistent würde ich gerne den MQTT Broker aktivieren.
Geht das dann auch wenn ich den TSUN Proxy auf dem NAS laufen lasse ?!

**Reply by s-allius** [reply, marked as answer]
Ja klar sollte dies gehen. Im Config file des Proxies musst du nur die IP-Adresse oder den DNS Namen deines MQTT Brokers angeben. Wenn beide im selben IP-Netz sind, sollte es direkt funktionieren. 

Wenn du den Rechner mit Home-Assistant nicht immer durchlaufen lässt, wäre es vermutlich sinnvoll, den MQTT Broker doch auf der NAS dauerhaft laufen zu lassen. Es gibt QoS Klassen im MQTT Protokoll, die Informationen nachhalten, sodass Home-Assistant diese auch später erhalten kann, zum Beispiel wenn du Home-Assistant nur bei Bedarf startest. Derzeit nutze ich dies nicht und ich gehe davon aus, dass der Proxy, MQTT Broker und Home Assistant immer laufen

**Comment by megabyte0469** [comment]
Hallo Stefan,
ist schon eine Zeit her. Leider läuft die Integration nicht mehr. Hatte einiges geändert im Heimnetz und nun leider die ganze Installation nicht mehr zum laufen bekommen. Du hattest ja geschrieben ein Installationsanleitung zu machen. Gibt es diese schon. Aktuell läuft Homeassistant hier auf einen extra Rechner 24h (NUC I3) und parallel hab ich noch ein NAS laufen. Ich hatte deine Integration ja auf dem NAS laufen und der MQTT Broker ist auf dem Homeassistant Rechner bereits wegen einer anderen Integration aktiv. Leider hat Synology ja die ursprüngliche Docker Integration komplett auf dem NAS geändert. Nun stehe ich leider etwas auf dem Kriegsfuss damit....

**Comment by s-allius** [comment]
Hallo megabyte0469,

eine Anleitung habe ich noch nicht erstellt. Vielleicht magst du mit einer starten?

Auf den NAS Systemen versuchen die Hersteller, immer eine "einfache" Oberfläche für Docker zu integrieren, die sie dann regelmäßig inkompatibel "verbessern". Nicht nur Synology macht dies...   
Daher installiere als erstes immer die Container-Verwaltung ´Portainer´. Dies ist dann der einzige Container den ich direkt auf der NAS aufsetze. Alles weiter mache ich dann in ´Portainer`.
Du kannst natürlich auch den TSUN Proxy auf dem NUC installieren. 

Was hast du den für ein Problem genau? Hast du schon die Prüfungen aus meiner ersten Antwort oben probiert?

Gruß Stefan

**Reply by 0009fe3** [reply]
Hello Stefan! Thank you for sharing this cool project with us! I am completely new to Docker and would also like to run the proxy on a Synology DiskStation. Unfortunately, I am currently not able to do it. The config.example.toml is not created. I have checked and adjusted the UID. I assume that I am doing something wrong with the volumes. If I manage to get the proxy running with your help, I would like to contribute and start documentation for the installation on a Synology NAS. Do you have a tip for me?

Thank you in advance!

![screenshot_volume_config](https://github.com/s-allius/tsun-gen3-proxy/assets/167771230/08f95153-1db7-41e2-ba55-d294d7687d9c)
![screenshot_volume_config](https://github.com/s-allius/tsun-gen3-proxy/assets/167771230/c154f5c5-8dcc-4400-836b-802be91fbbf6)
![screenshot_log](https://github.com/s-allius/tsun-gen3-proxy/assets/167771230/1779d747-2c40-4f94-afa5-7f1df7afe378)

**Comment by d4n7e1** [comment]
just download the .toml in app/config in this repository and upload it to your diskstation.

**Reply by 0009fe3** [reply]
I tried to upload the file, but nothing changed

![screenshot_log2](https://github.com/s-allius/tsun-gen3-proxy/assets/167771230/76a6cd11-94da-4d0a-bb8f-da3b59bc7f07)
![screenshot_config_uploaded](https://github.com/s-allius/tsun-gen3-proxy/assets/167771230/4bfcf368-fb26-445e-a2df-9ff64fcd7b19)

**Comment by s-allius** [comment]
I think this line shows the problem:
```
2024-05-29 13:10:55 ERROR I root | Confia.read: (Errno 21 No such file or directorv: 'confia/confia.toml
```
The proxy didn't know the base path, where to store the files. Since you use portianer and hopefully the `stacks` in it, you can easily fix this. In the `docker-compose.yaml` is a mapping for the files:
```
    volumes:
      - ${PROJECT_DIR}./tsun-proxy/log:/home/tsun-proxy/log 
      - ${PROJECT_DIR}./tsun-proxy/config:/home/tsun-proxy/config 
```
So to add the base path you should define the environment variable `PROJECR_DIR`, with a value like `/volume2/docker/`. So on your side the paths are a little bit different and you should adapt this.

If you don't use the `stacks` yet, I can only recommend it to you. Start with adding a new stack and copying the `docker-compose.yaml`into the editor field in your portioner. Then define the environment variables `PROJECT_DIR` and `UID`.

If that didn't work for you, pls don't hesitate to ask here again.

**Reply by s-allius** [reply]
I have one more idea to debug your problem. look in portainer under inspect/mounts, there you find your actual mapping. On my side it locks like
```
Mounts
0
Destination /home/tsun-proxy/config
Mode rw
Propagation rprivate
RW true
Source /volume2/docker/tsun-proxy/config
Type bind
1 
```
Pls check if your source path is correct.

**Reply by fff5ee** [reply]
Hi S-allius! Thanks for taking the time to respond to my message. Going through your answer, I realized that it would be better for me to start over from the beginning and learn some basics. :)  Portainer was installed but not in use. So, I dove a little bit into it, created all the folders needed on my Synology NAS, set the permissions for the Docker user, and picked up the UID of the user.

I then used a Portainer Stack and your compose file to create all the needed resources. The only thing I changed was the network parameter External = true. I removed it, so that Docker could create the network for me.

I think my main problem before was that the path for the environment variables PROJECT_DIR and UID was not set. So, you were completely right.

So right now, I have:

- A running TSUN-Proxy container
         - A DNS Server on my Synology which gets deployed to all clients via DHCP 
                           - With a Host A entry for logger.talent-monitoring.com, that resolves to the Docker Host (Synology NAS) -> Tested via 
                             nslookup

- A running MQTT Server which allows anonymous logon (which cannot be tested from my local client, because of the fact that the port is only exposed in the isolated network)

- A running Home Assistant Docker container, which I can access from my local client.
            - With the MQTT integration installed and successfully added the MQTT Broker with the MQTT Container IP

And finally, I can see some devices in Home Assistant, I was so happy about that fact :D

But I think there is still some problem, because every sensor is showing the value 0.

Do you have an idea where I could start troubleshooting?

Thank you so much!

![image](https://github.com/s-allius/tsun-gen3-proxy/assets/11533106/ca256771-c75c-4808-8584-fd6d0606426d)

**Reply by s-allius** [reply]
Congratulations.
The fact that most of the counters are zero is OK, because they are error counters. Only the Sensor "Active Inverter Connection" should increase to one.
More devices will be added to HA wenn the inverter start a connection for the first time. So you may wait for sunset.
But you can test this easily test the connection at night with you browser. Try this url:
http://<ip-of-your-NAS>:5005/

It is important that you use http and not https. The browser can't load a page and seems to hang. That's OK.
Now check the containter log of the proxy. you should see something like

```log
'2024-06-03 23:48:22  INFO | root | [G3] Accept connection from ('172.28.1.62', 34436)'
```

Then everything is fine. When on sunset your inverter will not appear, you should check your inverter settings in the config.toml file. You probably will have a mistake with the serial number of your inverter.

Regards Stefan

**Reply by fff5ee** [reply]
Thank you very much! You’re completely right. It was the first thing I did this morning to check Home Assistant, and the values are there and can be shown in the energy dashboard! That’s so cool. Thanks for your effort in this project!

Since you have a very similar setup to mine, can I ask you a question about MQTT even though this is not your code? As I mentioned earlier, I used your stack to deploy everything.

When I look in the MQTT logs, I see the following warning:

1717441584: Warning: File /mosquitto/data//mosquitto.db has world readable permissions. Future versions will refuse to load this file. To fix this, use chmod 0700 /mosquitto/data//mosquitto.db`.

I have tried to search for a resolution on the web, but didn’t find something that works. I also tried to use the command chmod 0700 /mosquitto/data//mosquitto.db in the Docker interactive session, but no change at all.

I think I cannot set a username and password because of this issue right now. If you do not want to invest time in this and only want to talk about your project here, I can totally understand that!

Regards

**Reply by s-allius** [reply]
Yes I had this warning before too. To fix this you must login as Admin via SSH. Then goes into the directory of you docker.
Might be something like:
```bash
$ cd /volume1/docker/mosquitto
$ ls -al
total 0
drwxrwxrwx+ 1 Admin_jfjaslje users  20 Sep  9  2023 .
drwxrwxrwx+ 1 root         root  202 Jun  3 20:15 ..
drwxrwxrwx+ 1         1883  1883  50 Oct  4  2023 config
drwx------  1         1883  1883  24 Jun  4 22:55 data

$ sudo ls data -al
total 24
drwx------  1         1883  1883    24 Jun  4 22:55 .
drwxr-xr-x+ 1 Admin_jfjaslje users  20 Sep  9  2023 .
-rw-------  1         1883  1883 24082 Jun  4 22:55 mosquitto.db
```
There you can adapt the rights with chmod for the `data`dir and  the file mosquitto.db. Above you can see, that I only have rights for the admin role for them.

**Comment by s-allius** [comment]
@fff5ee 
Do you have some notes or a description how you did it? 
Than I can start with a first version of a How Too in the wiki.

You can also fork the wiki and write it by your own. Than I can merge you notes as a pull request into this project.

**Reply by fff5ee** [reply]
@s-allius:
Not now, but I plan to invest some time, hopefully this weekend, starting to describe the whole setup. 

Here's a quick overview of the steps I plan to write down:

1. Create needed folders on NAS
2. Create a user for Docker
3. Install Portainer using the Synology GUI
4. Deploy software using YAML, including explaining a little bit about the environment variables, UID, and network parts, which were the most challenging for me
5. Modify the MQTT Broker to listen to anonymous connections and hopefully (if I get it running on my side) secure it with a username/password. Explain the needed changes when you want to use the MQTT broker from outside of the Docker network, if you want to use it for an AMSI Smart Meter Reader for example (Expose vs Port)
6. Modify the config file from your proxy
7. Install MQTT integration in Home Assistant and integrate the entities
8. Add it to the energy dashboard

I know following the steps would only work with the same setup as me, but I think it could still be useful to see a "full deployment" way.

What do you think about that approach?

**Reply by s-allius** [reply]
Yes, I think that's a good approach and it's also completely okay if you describe your example exactly. We can then generalize this piece by piece.
I already have a question in another forum about how to install the proxy on a synology. I think it will help many.
